pkg/tls/tls.go

```diff
@@ -0,0 +1,238 @@
+package tls
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	// Need to embed the default config into the library
+	_ "embed"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"path"
+	"time"
+)
+
+var (
+	privateNodeNames = []string{
+		"full_node",
+		"wallet",
+		"farmer",
+		"harvester",
+		"timelord",
+		"crawler",
+		"data_layer",
+		"daemon",
+	}
+	publicNodeNames = []string{
+		"full_node",
+		"wallet",
+		"farmer",
+		"introducer",
+		"timelord",
+		"data_layer",
+	}
+
+	//go:embed chia_ca.crt
+	chiaCACrtBytes []byte
+
+	//go:embed chia_ca.key
+	chiaCAKeyBytes []byte
+)
+
+// GenerateAllCerts generates the full set of required certs for chia blockchain
+func GenerateAllCerts(outDir string) error {
+	// First, ensure that all output directories exist
+	allNodes := append(privateNodeNames, publicNodeNames...)
+	for _, subdir := range append(allNodes, "ca") {
+		err := os.MkdirAll(path.Join(outDir, subdir), 0700)
+		if err != nil {
+			return fmt.Errorf("error making output directory for certs: %w", err)
+		}
+	}
+
+	// Next, copy the chia_ca cert/key
+	err := os.WriteFile(path.Join(outDir, "ca", "chia_ca.crt"), chiaCACrtBytes, 0600)
+	if err != nil {
+		return fmt.Errorf("error copying chia_ca.crt: %w", err)
+	}
+	err = os.WriteFile(path.Join(outDir, "ca", "chia_ca.key"), chiaCAKeyBytes, 0600)
+	if err != nil {
+		return fmt.Errorf("error copying chia_ca.key: %w", err)
+	}
+
+	chiaCACert, err := ParsePemCertificate(chiaCACrtBytes)
+	if err != nil {
+		return fmt.Errorf("error parsing chia_ca.crt")
+	}
+
+	chiaCAKey, err := ParsePemKey(chiaCAKeyBytes)
+	if err != nil {
+		return fmt.Errorf("error parsing chia_ca.key")
+	}
+
+	privateCACertBytes, privateCAKeyBytes, err := GenerateNewCA(path.Join(outDir, "ca", "private_ca"))
+	if err != nil {
+		return fmt.Errorf("error creating private ca pair: %w", err)
+	}
+	privateCACert, err := ParsePemCertificate(privateCACertBytes)
+	if err != nil {
+		return fmt.Errorf("error parsing generated private_ca.crt: %w", err)
+	}
+	privateCAKey, err := ParsePemKey(privateCAKeyBytes)
+	if err != nil {
+		return fmt.Errorf("error parsing generated private_ca.key: %w", err)
+	}
+
+	for _, node := range publicNodeNames {
+		_, _, err = GenerateCASignedCert(chiaCACert, chiaCAKey, path.Join(outDir, node, fmt.Sprintf("public_%s", node)))
+		if err != nil {
+			return fmt.Errorf("error generating public pair for %s: %w", node, err)
+		}
+	}
+
+	for _, node := range privateNodeNames {
+		_, _, err = GenerateCASignedCert(privateCACert, privateCAKey, path.Join(outDir, node, fmt.Sprintf("private_%s", node)))
+		if err != nil {
+			return fmt.Errorf("error generating private pair for %s: %w", node, err)
+		}
+	}
+
+	return nil
+}
+
+// ParsePemCertificate parses a certificate
+func ParsePemCertificate(certPem []byte) (*x509.Certificate, error) {
+	// Load CA certificate
+	caCertBlock, _ := pem.Decode(certPem)
```

Comment:
What's the other return for `pem.Decode`? Does it make sense to not check it?

Reply:
You probably get similar error handling from the if statement just below though, so it _might_ be fine to ignore the error. However it may add context?